File: android/app/src/main/kotlin/com/nomadguy/noctra/lyricromanizer/engines/CyrillicEngine.kt
package com.nomadguy.noctra.lyricromanizer.engines

import com.nomadguy.noctra.lyricromanizer.RomanizeEngineContext

/** Cyrillic to Latin romanizer with Ukrainian auto-detection. */
class CyrillicEngine {

    private val ukraineRe = Regex("[іїєґ]", RegexOption.IGNORE_CASE)

    private val ruMap = mapOf(
        'а' to "a", 'б' to "b", 'в' to "v", 'г' to "g", 'д' to "d",
        'е' to "e", 'ё' to "yo", 'ж' to "zh", 'з' to "z", 'и' to "i",
        'й' to "y", 'к' to "k", 'л' to "l", 'м' to "m", 'н' to "n",
        'о' to "o", 'п' to "p", 'р' to "r", 'с' to "s", 'т' to "t",
        'у' to "u", 'ф' to "f", 'х' to "kh", 'ц' to "ts", 'ч' to "ch",
        'ш' to "sh", 'щ' to "shch", 'ъ' to "", 'ы' to "y", 'ь' to "",
        'э' to "e", 'ю' to "yu", 'я' to "ya",
        'А' to "A", 'Б' to "B", 'В' to "V", 'Г' to "G", 'Д' to "D",
        'Е' to "E", 'Ё' to "Yo", 'Ж' to "Zh", 'З' to "Z", 'И' to "I",
        'Й' to "Y", 'К' to "K", 'Л' to "L", 'М' to "M", 'Н' to "N",
        'О' to "O", 'П' to "P", 'Р' to "R", 'С' to "S", 'Т' to "T",
        'У' to "U", 'Ф' to "F", 'Х' to "Kh", 'Ц' to "Ts", 'Ч' to "Ch",
        'Ш' to "Sh", 'Щ' to "Shch", 'Ъ' to "", 'Ы' to "Y", 'Ь' to "",
        'Э' to "E", 'Ю' to "Yu", 'Я' to "Ya"
    )

    private val ukMap = mapOf(
        'і' to "i", 'ї' to "yi", 'є' to "ye", 'ґ' to "g",
        'І' to "I", 'Ї' to "Yi", 'Є' to "Ye", 'Ґ' to "G"
    )

    /** Selects the Ukrainian preset for Ukrainian-specific characters. */
    fun selectPreset(line: String): String =
        if (ukraineRe.containsMatchIn(line)) "uk" else "ru"

    suspend fun romanize(line: String, context: RomanizeEngineContext): String {
        val preset = selectPreset(line)
        val buf = StringBuilder()
        for (ch in line) {
            if (preset == "uk" && ukMap.containsKey(ch)) {
                buf.append(ukMap[ch])
            } else {
                buf.append(ruMap[ch] ?: ch)
            }
        }
        return buf.toString()
    }
}

File: lib/services/lyric_romanizer/engines/cyrillic_engine.dart
/// Cyrillic to Latin romanization engine.
///
/// Auto-detects Ukrainian vs Russian per line using the presence of
/// Ukrainian-specific characters (і, ї, є, ґ).

library;
import '../lyric_romanizer_types.dart';

class CyrillicRomanizer {
  const CyrillicRomanizer();

  /// Ukrainian-specific characters that trigger the Ukrainian preset.
  static final RegExp _ukrainePresetRe = RegExp(r'[іїєґ]', caseSensitive: false);

  /// Detects whether the line contains Ukrainian-specific characters.
  static String selectPreset(String line) =>
      _ukrainePresetRe.hasMatch(line) ? 'uk' : 'ru';

  static const Map<String, String> _ruMap = {
    'а': 'a', 'б': 'b', 'в': 'v', 'г': 'g', 'д': 'd', 'е': 'e',
    'ё': 'yo', 'ж': 'zh', 'з': 'z', 'и': 'i', 'й': 'y', 'к': 'k',
    'л': 'l', 'м': 'm', 'н': 'n', 'о': 'o', 'п': 'p', 'р': 'r',
    'с': 's', 'т': 't', 'у': 'u', 'ф': 'f', 'х': 'kh', 'ц': 'ts',
    'ч': 'ch', 'ш': 'sh', 'щ': 'shch', 'ъ': '', 'ы': 'y', 'ь': '',
    'э': 'e', 'ю': 'yu', 'я': 'ya',
    'А': 'A', 'Б': 'B', 'В': 'V', 'Г': 'G', 'Д': 'D', 'Е': 'E',
    'Ё': 'Yo', 'Ж': 'Zh', 'З': 'Z', 'И': 'I', 'Й': 'Y', 'К': 'K',
    'Л': 'L', 'М': 'M', 'Н': 'N', 'О': 'O', 'П': 'P', 'Р': 'R',
    'С': 'S', 'Т': 'T', 'У': 'U', 'Ф': 'F', 'Х': 'Kh', 'Ц': 'Ts',
    'Ч': 'Ch', 'Ш': 'Sh', 'Щ': 'Shch', 'Ъ': '', 'Ы': 'Y', 'Ь': '',
    'Э': 'E', 'Ю': 'Yu', 'Я': 'Ya',
  };

  static const Map<String, String> _ukMap = {
    'і': 'i', 'ї': 'yi', 'є': 'ye', 'ґ': 'g',
    // Ukrainian и maps to 'y' (differs from Russian и → 'i')
    'и': 'y', 'И': 'Y',
    'І': 'I', 'Ї': 'Yi', 'Є': 'Ye', 'Ґ': 'G',
  };

  Future<String> romanize(String line, RomanizeEngineContext context) async {
    final preset = selectPreset(line);
    final buf = StringBuffer();
    for (final ch in line.split('')) {
      if (preset == 'uk' && _ukMap.containsKey(ch)) {
        buf.write(_ukMap[ch]);
      } else {
        buf.write(_ruMap[ch] ?? ch);
      }
    }
    return buf.toString();
  }
}
